!!! BUGFIX: Refactor FusionService API & And strict FusionCodeCollection in FusionParser

mhsdesign opened the pull request:
Improve the API of the fusion parser and introduce FusionSourceCodeFactory to create fusion auto includes.

In the past the auto includes were added to the main Root.Fusion from the current website. With the addition of the Fusion Ast Cache in Development context this caused issues when site packages included the Root.fusion of other site packages as the same cache entry was used for the file wether it was included as main or not.

The change introduces the parseFromSource() Method that accepts a FusionSourceCodeCollection DTO to the Fusion parser and deprecates the parse() method.

resolves: #3835
resolves: #3834

**Upgrade Instructions**

While this change does not alter the api it deprecates the method `Neos\Fusion\Core\Parser::parse()`
You should update it to `Neos\Fusion\Core\Parser::parseFromSource()`. The external api of the `Neos\Neos\Domain\Service\FusionService` was not changed. In the unlikely case that you extended this service and used internal methods you probably will have to adjust your code. 

**Review Instructions**

As discussed in https://github.com/neos/neos-development-collection/issues/3835#issuecomment-1181840848 a `FusionCodeCollection` can now be provided as argument to `$fusionParser-> parseFromSource()`. This enforces that not accidentally differing `$sourceCode` and `$contextPathAndFilename` are given, which confuses the cache.

Also we all love to cure our [primitive (type) obsession](https://refactoring.guru/smells/primitive-obsession) :smile: 

The change is made backwards compatible, so one can still provide the primitive types.

To achieve the current dangerous behavior, provide differing `$sourceCode` and `$contextPathAndFilename`, use the `->parse()` API with primitive types, or use the escape hatch left open via `FusionCode::fromDangerousPotentiallyDifferingSourceCodeAndContextPath()`, in order to make AOP / Fusion preprocessing still possible - but with caution ;)


**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX 3835 - FusionParser stricter $contextPathAndFileName handling via FusionCodeCollection

mhsdesign committed:
BUGFIX 3835 - FusionParser pass 'FusionCode' Dto to ObjectTreeParser

mhsdesign committed:
BUGFIX 3835 - FusionService allow legacy relative 'appendFusionIncludes'

mhsdesign commented:
**Original description**

As discussed in https://github.com/neos/neos-development-collection/issues/3835#issuecomment-1181840848 a `FusionCodeCollection` can now be provided as argument to `$fusionParser->parse()`. This enforces that not accidentally differing `$sourceCode` and `$contextPathAndFilename` are given, which confuses the cache.

Also we all love to cure our [primitive (type) obsession](https://refactoring.guru/smells/primitive-obsession) :smile: 

The change is made backwards compatible, so one can still provide the primitive types.

To achieve the current dangerous behavior, provide differing `$sourceCode` and `$contextPathAndFilename`, use the `->parse()` API with primitive types, or use the escape hatch left open via `FusionCode::fromDangerousPotentiallyDifferingSourceCodeAndContextPath()`, in order to make AOP / Fusion preprocessing still possible - but with caution ;)

resolves: #3835
resolves: #3834

## upgrade-instructions
`Neos\Fusion\Core\Parser::parse()` is deprecated.
You should update it to `Neos\Fusion\Core\Parser::parseFrom()`

if you extend `Neos\Neos\Domain\Service\FusionService` you most likely need to adjust it if you used any internals.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions


--------------


The Neos.Neos FusionService Test is a bit odd as it uses the already odd api `setAppendFusionIncludes` with relative fusionFiles https://github.com/neos/neos-development-collection/blob/dd64e385aecb668d609b8a5e028c64541898282d/Neos.Neos/Tests/Functional/Fusion/RenderingTest.php#L362

this is now again "supported" via https://github.com/neos/neos-development-collection/pull/3839/commits/52f4ee1b19f2159e101ed2e17eb88b155bd217df

Imo the test should not use this wierd api - and prependFusionIncludes and appendFusionIncludes should be removed, as i dont see it being used, and cannot find a use for it - except the one in the testcase.

mhsdesign commented on Neos.Neos/Classes/Domain/Service/FusionService.php:
```diff
@@ -159,33 +161,23 @@ protected function getSiteForSiteNode(TraversableNodeInterface $siteNode)
      * @throws \Neos\Neos\Domain\Exception
      */
-    protected function generateNodeTypeDefinitions()
+    protected function generateNodeTypeDefinitions(): ?string
```
should return a dto too

mficzel commented on Neos.Fusion/Classes/Core/FusionCode.php:
```diff
@@ -0,0 +1,61 @@
+use Neos\Flow\Annotations as Flow;
+
+#[Flow\Proxy(false)]
+class FusionCode
```
Find better naming: FusionCodeRefernce, RawFusionCode, FusionParserInput. ...

mficzel commented on Neos.Fusion/Classes/Core/Parser.php:
```diff
@@ -50,20 +50,33 @@ class Parser implements ParserInterface
      * @api
      */
-    public function parse(string $sourceCode, ?string $contextPathAndFilename = null, array $mergedArrayTreeUntilNow = []): array
+    public function parse(string|FusionCodeCollection $sourceCode, ?string $contextPathAndFilename = null, array $mergedArrayTreeUntilNow = []): array
```
allow string|FusionCode|FusionCodeCollection $sourceCode,
mark deprecations with @deprecated

mficzel commented on Neos.Fusion/Classes/Core/Parser.php:
```diff
@@ -50,20 +50,33 @@ class Parser implements ParserInterface
      * @api
      */
-    public function parse(string $sourceCode, ?string $contextPathAndFilename = null, array $mergedArrayTreeUntilNow = []): array
+    public function parse(string|FusionCodeCollection $sourceCode, ?string $contextPathAndFilename = null, array $mergedArrayTreeUntilNow = []): array
```
allow string|FusionCode|FusionCodeCollection $sourceCode,
mark deprecations with @deprecated

mficzel replied:
Alternatively add new method `parseFrom` and deprecate the whole parse method

mhsdesign committed:
BUGFIX 3835 Adjust FusionCode to FusionSourceCode and introduce new `Parser::parseFrom`

`Parser::parse` is @deprecated

mhsdesign committed:
BUGFIX 3835 Adjust Fusion Tests to use new api `Parser::parseFrom`

mhsdesign commented:
is now even readier

kdambekalns commented on Neos.Fusion/Classes/Core/ParserInterface.php:
```diff
@@ -22,6 +22,17 @@
+     * @throws Fusion\Exception
+     * @api
+     */
+    public function parseFrom(FusionSourceCode|FusionSourceCodeCollection $sourceCode): array;
```
This makes it a breaking change, strictly speaking.

mhsdesign replied:
because a new method?

@mficzel and me decided to create a new method than to make the old `parse` method understand `FusionSourceCodeCollection` via union types as this makes things uneccessary complex

i think this is fine

kdambekalns replied:
Simply because it changes an interface marked as `@api` 😉

mhsdesign replied:
ahh i see xD

is this really important? otherwise i would add it to the interface later ... (with 9) -> or forget it ^^

kdambekalns replied:
Well… "strictly speaking". The rules would say, we must put a breaking change in the next major. The again, the `ParserInterface` is something I was surprised to see as `@api` and I really doubt someone already implemented their own. 🙃

If noone else objects, fine with me!

mhsdesign replied:
yes the `@api` seems odd -> as it cannot even be changed via Objects yaml as we use `new Parser()`

kdambekalns replied:
Then maybe drop it (the interface, not the `@api`), before someone notices? 🙈

mhsdesign replied:
^^
about removing the interface @mficzel said: https://github.com/neos/neos-development-collection/pull/3497#issuecomment-974801376

kdambekalns replied:
True… then again, `Parser` has only one public method, so that makes it clear, too. But of course we can leave the interface in and just drop the `@api`.

mhsdesign replied:
Fine with me either way. Was my gut feeling at first too.

Maybe @kitsunet can help us out?

kitsunet replied:
> Well… "strictly speaking". The rules would say, we must put a breaking change in the next major. The again, the `ParserInterface` is something I was surprised to see as `@api` and I really doubt someone already implemented their own. 🙃
> 
> If noone else objects, fine with me!

I'd say `@api` or not, it's an interface, which _could_ strictly speaking be replaced. Now I doubt anyone did, but that falls under famous last words. Do we need this change for anything really important like NOW? Otherwise I would prefer next major.

mhsdesign replied:
Nope the `ParserInterface` **CANNOT** be replaced via objects.yaml. Its only usage/reference is in beeing implemented by the `Parser`. It has **no use** than to delay us in the discussion ^^

We have 3 Solutions:

- add the new method to the `ParserInterface` in 9.0 (drop the `ParserInterface` in 9.0)
- drop the  `ParserInterface` now - as it has no use - and 0 breakiness - https://github.com/neos/neos-development-collection/pull/3839#discussion_r991112012
- add the new method to the `ParserInterface` now


im team "drop the  `ParserInterface` now"

any objections?

kitsunet replied:
Given that it's a useless interface I doubt anyone has really implemented it, so from my side I would be fine with this technically breaking change. I am SURE there is a project where someone imlpemented the interface but never checked if it works and it breaks, but weeeeeell......

mhsdesign replied:
just to be 100% clear this goes into 8.0 

If anyone objects, I will remove it with 8.2 8.2

mhsdesign replied:
I got lost in the little things - this is not what this PR should focus on - nothing will be changed about the `ParserInterface`

see https://github.com/neos/neos-development-collection/pull/3932 for a followup

kdambekalns commented on Neos.Fusion/Classes/Core/FusionSourceCodeIsInvalid.php:
```diff
@@ -0,0 +1,31 @@
+#[Flow\Proxy(false)]
+class FusionSourceCodeIsInvalid extends Exception
+{
+    public static function becauseTheSourceCodeIsEmpty(?string $contextPathAndFileName): self
+    {
+        return new self("The sourcecode of the FusionSourceCode must not be empty. $contextPathAndFileName", 1657963664);
+    }
+
+    public static function becauseTheFileNameIsNotReadable(string $attemptedFileName): self
+    {
+        return new self("Trying to read Fusion source code from file, but '$attemptedFileName' is not readable.", 1657963790);
+    }
```
I **hate** that whenever Doctrine throws an exception! Also, it makes the exception code no longer point to *the one place* the exception really comes from. Please, let us not introduce this "pattern".

mhsdesign replied:
fine with me ;)

mhsdesign replied:
@nezaniel just said that the new escr may or may not look entirely like this ^^

kdambekalns replied:
For me I have rarely seen the need for this. Most exceptional things happen in one place. Elsewhere something else goes wrong, so why would I add a static method for that?

I'd give in to the team only if this is introduced with the code being a method argument - because that unique ID has helped tremendously over the years.

mhsdesign replied:
yes in this case i did it only for fun - and because it looks cool 😎 - and i like this pattern.

i think this could easily grow into a big discussion and i agree on both sides (i have talked with some people already about this)

for this particular change, im fine with reverting.

kdambekalns replied:
I'd welcome that. We should discuss this separately.

mhsdesign replied:
okay lets move the discussion to https://discuss.neos.io/t/rfc-updated-default-code-style/5836?u=marc

kdambekalns commented on Neos.Fusion/Classes/Core/FusionSourceCode.php:
```diff
@@ -0,0 +1,61 @@
+        private ?string $contextPathAndFilename
+    ) {
+        if (trim($sourceCode) === '') {
+            throw FusionSourceCodeIsInvalid::becauseTheSourceCodeIsEmpty($this->contextPathAndFilename);
```
Please, no static "factories" for exceptions!

mhsdesign replied:
fine with me ;)

robertlemke replied:
Haven't seen such a way to instantiate an exception before. My gut feeling says: unnecessarily fancy / complicated. But maybe it's also just unfamiliar.

mhsdesign replied:
okay ;)
https://github.com/neos/neos-development-collection/pull/3839#discussion_r991122895

lets move the discussion to https://discuss.neos.io/t/rfc-updated-default-code-style/5836?u=marc

kdambekalns commented on Neos.Fusion/Classes/Core/FusionSourceCode.php:
```diff
@@ -0,0 +1,61 @@
+    {
+        $sourceCode = file_get_contents($fileName);
+        if ($sourceCode === false) {
+            throw FusionSourceCodeIsInvalid::becauseTheFileNameIsNotReadable($fileName);
```
Please, no static "factories" for exceptions!

mhsdesign replied:
fine with me ;)

robertlemke commented on Neos.Fusion/Classes/Core/FusionSourceCode.php:
```diff
@@ -0,0 +1,61 @@
+
+use Neos\Flow\Annotations as Flow;
+
+#[Flow\Proxy(false)]
```
Just for the record – did this class cause any compile error when proxy class building was enabled?

mhsdesign replied:
`new self()` doesnt work i think - and i dislike `static`

mhsdesign replied:
@fcool  adjusted with

3878a04382a5489c09dc71452e606485db8739aa

PRGfx commented:
> Imo the test should not use this wierd api - and prependFusionIncludes and appendFusionIncludes should be removed, as i dont see it being used, and cannot find a use for it - except the one in the testcase.

I actually used it once to ensure controller fusion exists before I expected fusion loaded trough a site-package setup. I would imagine that's what these options are for and it makes sense to me. However in practice it is a bit difficult to use, as an aspect caches this fusion tree "globally" based on the first request to build it, i.e. using the fusion service in another context will re-use the same cached fusion tree..

mhsdesign commented on Neos.Fusion/Classes/Core/FusionSourceCode.php:
```diff
@@ -0,0 +1,61 @@
+
+    public static function fromFile(string $fileName): self
+    {
+        $sourceCode = file_get_contents($fileName);
```
should this `file_get_contents` happen lazily?

eg - it might happen that the `FusionSourceCode` is already cached and then we would not need its contents

mhsdesign committed:
BUGFIX 3835 use normal Fusion\Exception instead of  FusionSourceCodeIsInvalid with factory methods

mhsdesign commented on Neos.Fusion/Classes/Core/FusionSourceCode.php:
```diff
@@ -0,0 +1,62 @@
+        private string $sourceCode,
+        private ?string $contextPathAndFilename
+    ) {
+        if (trim($sourceCode) === '') {
+            throw new Fusion\Exception("The sourcecode of the FusionSourceCode must not be empty. $this->contextPathAndFilename", 1657963664);
+        }
```
We cant do that, as its currently allowed to have an empty `Root.fusion`

that would be breaking so im reverting this.

![image](https://user-images.githubusercontent.com/85400359/197333973-9408e771-ae5f-4478-a674-7858f29d563a.png)

mhsdesign committed:
BUGFIX 3835 remove accidental breaking change

if (trim($sourceCode) === '') throw Error

We cant do that, as its currently allowed to have an empty Root.fusion

https://github.com/neos/neos-development-collection/pull/3839/files#r1002435846

mhsdesign commented:
## ~last remaining discussion~

~should the completely unused ParserInterface be removed with this pr retroactiv for Neos 8.0 and 8.1~

~see https://github.com/neos/neos-development-collection/pull/3839#discussion_r1002434260~

update1
~removed the interface now for 8.0~

update2
I got lost in the little things - this is not what this PR should focus on - nothing will be changed about the `ParserInterface`

mhsdesign committed:
BUGFIX 3835 adjust php-doc

mhsdesign commented:
removed the interface now for 8.0

mhsdesign commented:
this is now ready for review ;)

kdambekalns commented on Neos.Fusion/Classes/Core/Parser.php:
```diff
@@ -83,7 +83,7 @@ public function parseFrom(FusionSourceCode|FusionSourceCodeCollection $sourceCod
      * @return array The merged array tree for the Fusion runtime, generated from the source code
      * @throws Fusion\Exception
-     * @deprecated with Neos 8.0
+     * @deprecated with Neos 8.0 use {@link parseFrom} will be removed with Neos 9.0
```
```suggestion
     * @deprecated with Neos 8.0 – will be removed with Neos 9.0, use {@link parseFrom} instead
```

kdambekalns commented on Neos.Fusion/Classes/Core/ParserInterface.php:
```diff
@@ -18,21 +18,11 @@
 /**
  * Contract for a Fusion parser
  *
+ * @deprecated with Neos 8.0 since the interface is unused will be removed with Neos 9.0
```
```suggestion
 * @deprecated with Neos 8.0 since the interface is unused, will be removed with Neos 9.0
```

kdambekalns commented on Neos.Fusion/Classes/Core/ParserInterface.php:
```diff
@@ -42,8 +32,8 @@ public function parseFrom(FusionSourceCode|FusionSourceCodeCollection $sourceCod
      * @param array $mergedArrayTreeUntilNow Used internally for keeping track of the built merged array tree
      * @return array The merged array tree for the Fusion runtime, generated from the source code
      * @throws Fusion\Exception
+     * @deprecated with Neos 8.0 use {@link Parser::parseFrom} will be removed with Neos 9.0
```
```suggestion
     * @deprecated with Neos 8.0 – will be removed with Neos 9.0, use {@link Parser::parseFrom} instead
```

kdambekalns reviewed (commented):
Some doc improvements…

mhsdesign commented on Neos.Neos/Classes/Domain/Service/FusionService.php:
```diff
@@ -136,57 +137,40 @@ public function createRuntime(TraversableNodeInterface $currentSiteNode, Control
-        $mergedFusionCode .= $siteRootFusionCode;
-        $mergedFusionCode .= $this->getFusionIncludes($this->appendFusionIncludes);
 
-        return $this->fusionParser->parse($mergedFusionCode, $siteRootFusionPathAndFilename);
```
removing here the use of a possible non existing file $siteRootFusionPathAndFilename
fixes
https://github.com/neos/neos-development-collection/pull/3834

mhsdesign commented on Neos.Neos/Classes/Domain/Service/FusionService.php:
```diff
@@ -136,57 +137,40 @@ public function createRuntime(TraversableNodeInterface $currentSiteNode, Control
+        return $this->fusionParser->parseFrom(
+            FusionSourceCodeCollection::fromArray(array_filter([
+                ...$this->generateNodeTypeDefinitions(),
+                ...$this->prepareAutoIncludeFusion(),
```
The autoincludes are now separate new entry points instead of concatenated fusion code with $siteRootFusionPathAndFilename as wrong context file 

fixes
https://github.com/neos/neos-development-collection/issues/3835

mhsdesign reviewed (commented):
This looks in total now more like a refactoring for fun - but there are some bugfixes hidden inside:

kdambekalns reviewed (approved):
by reading

mhsdesign commented on Neos.Fusion/Classes/Core/ParserInterface.php:
```diff
@@ -18,6 +18,7 @@
 /**
  * Contract for a Fusion parser
  *
+ * @deprecated with Neos 8.0 since the interface is unused, will be removed with Neos 9.0
```
```suggestion
 * @deprecated with Neos 8.0 since the interface is unused, will be removed with Neos 8.2
```

mhsdesign commented on Neos.Fusion/Classes/Core/ParserInterface.php:
```diff
@@ -31,6 +32,7 @@ interface ParserInterface
      * @param array $mergedArrayTreeUntilNow Used internally for keeping track of the built merged array tree
      * @return array The merged array tree for the Fusion runtime, generated from the source code
      * @throws Fusion\Exception
+     * @deprecated with Neos 8.0 – will be removed with Neos 9.0, use {@link Parser::parseFrom} instead
```
```suggestion
     * @deprecated with Neos 8.0 - use {@link Parser::parseFrom} instead
```

mhsdesign committed:
BUGFIX 3835 use `static` instead of `self` with `#[Flow\Proxy(false)]`

mhsdesign committed:
BUGFIX 3835 cleanup legacy mapping in Neos\Fusion\Core\Parser::parse

mficzel commented on Neos.Neos/Classes/Domain/Service/FusionService.php:
```diff
@@ -136,57 +137,40 @@ public function createRuntime(TraversableNodeInterface $currentSiteNode, Control
+                ...$this->prepareFusionIncludes($this->appendFusionIncludes, $siteRootFusionPathAndFilename),
+            ]))
+        );
     }
```
This array filtering and unpacking is not that easy to read. How about something like:
```
new FusionSourceCodeCollection (
     $this->generateNodeTypeDefinitions(),
     $this->prepareAutoIncludeFusion(),
     $this->prepareFusionIncludes($this->prependFusionIncludes, $siteRootFusionPathAndFilename),
     is_readable($siteRootFusionPathAndFilename) ? FusionSourceCode::fromFile($siteRootFusionPathAndFilename) : FusionSourceCode::empty(),
     $this->prepareFusionIncludes($this->appendFusionIncludes, $siteRootFusionPathAndFilename),
);
```
this would require to change the constuctor of the FusionSourceCodeCollection to
```
class FusionSourceCodeCollection implements \IteratorAggregate, \Countable
{
    public function __construct(FusionSourceCode|FusionSourceCodeCollection ...$fusionCodeCollection)
```
and do the collection unpacking in there

also `FusionSourceCode::empty()` would remove the need for the filtering of null valies

mficzel commented on Neos.Fusion/Tests/Unit/Core/ParserTest.php:
```diff
@@ -63,7 +64,7 @@ public function parserCorrectlyParsesFixture01()
         ];
 
-        $actualParseTree = $this->parser->parse($sourceCode);
+        $actualParseTree = $this->parser->parseFrom(\Neos\Fusion\Core\FusionSourceCode::fromString($sourceCode));
```
I think at least some of the old parse* tests should stay there for now so we know that those do still work as before until 9.0 ... otherwise we may break stuff for packages like monocle without noticing

mhsdesign replied:
yes youre right ;) ill add an extra test file which can later be removed ;)

mficzel commented on Neos.Fusion/Classes/View/FusionView.php:
```diff
@@ -184,17 +185,17 @@ protected function loadFusion()
             }
         }
-        return $parsedFusion;
+        return $this->fusionParser->parseFrom(FusionSourceCodeCollection::fromArray($fusionCodeCollection));
```
imho could be       
```
return $this->fusionParser->parseFrom(new FusionSourceCodeCollection( ...$fusionFiles ));

mficzel commented on Neos.Neos/Classes/Domain/Service/FusionService.php:
```diff
@@ -305,4 +268,18 @@ public function setAppendFusionIncludes(array $appendFusionIncludes)
         $this->appendFusionIncludes = $appendFusionIncludes;
     }
+
+    private function prepareFusionIncludes(array $fusionIncludes, string $filePathForRelativeResolves): FusionSourceCodeCollection
```
Please make this `protected` so other (monocle) packages may use it.

mhsdesign replied:
![image](https://user-images.githubusercontent.com/85400359/200027849-53448993-e2e8-4e25-8113-000afd7b7b7b.png)

mficzel replied:
Having a custom FusionService that extends the original one is not evil and we have at least one valid use case for that.

mhsdesign replied:
yes sorry i was not trying to be mean. Its just that our whole `FusionService` is broken.

My masterplan is to remove the whole `FusionService` at some point and so i was trying to actually limit the api surface.

mficzel replied:
Actually it would be fine to declare the scope of those methods whatever you like if you add something like this:

```
public function getMergedFusionObjectTreeForSitePackage(FlowPackageInterface $sitePackage)
```

In which case monocle would not extend the service but simply use it.

mhsdesign replied:
a little lobbyism never did any harm ^^

I really like this api and clear approach!

mhsdesign replied:
after thinking A LOT and also consulting you @nezaniel and @kitsunet i found a clean api for every usecase:

```php
$fusionConfiguration = $this->fusionParser->parseFromSource(
    $this->fusionSourceCodeFactory->createFromNodeTypeDefinitions()
        ->union(
            $this->fusionSourceCodeFactory->createFromAutoIncludes()
        )
        ->union(
            FusionSourceCodeCollection::tryFromPackageFusionRoot("Vendor.Site")
        )
);
```

i hope you like it ;)

mficzel reviewed (changes requested):
Generally fine and it works on the Neos side. It breaks Monocle but i the needed adjustments are minimal. 

Suggested some adjustments to avoid unpacking and filtering in the fusion service.

kitsunet commented on Neos.Fusion/Classes/Core/Parser.php:
```diff
@@ -46,6 +46,34 @@ class Parser implements ParserInterface
+     * @throws Fusion\Exception
+     * @api
+     */
+    public function parseFrom(FusionSourceCode|FusionSourceCodeCollection $sourceCode): array
```
you _could_ introduce a common interface for the two that allows iterating (for both, just one entry for the single thing).

mhsdesign replied:
yes why not ;)

mhsdesign replied:
I find it a challeging idea ... I don't think there can be a common interface ...
I just created for fun an "Eierlegendewollmilchsau" dto:

```php
class Dto implements IteratorAggregate
{
  private function __construct(private string $name, private int $value, private array $dtos)
  {
  }

  public static function create(string $name, int $value): Dto
  {
    return new self($name, $value);
  }

  public static function fromDtos(Dto ...$dtos): Dto {
    return new self(dtos: $dtos);
  }

  public function getName(): string
  {
     return $this->name ?? $this->dtos[0]->name;
  }

  public function getValue(): string
  {
     return $this->value = $this->dtos[0]->value;
  }

  public function getIterator(): ArrayIterator
  {
    return new ArrayIterator($this->dtos ?? [$this]);
  }
}
```

this seems like the worst idea one can have :D as you'd never know if you hold a real iterable or the final values ^^ Also there is no guarantee that only either name and value or dtos is set as property but it could have a wrong state which is a code smell per se. Also it violates SRP ^^

kitsunet replied:
Yeah obviously not a single object, but if you implement an recursive iterator across both and make that a common interface, so that you can just take whatever object you have at hand and do `foreach ($fusionCollection as $name => $value)` or something and it gives you just recursively all `FusionSourceCode` objects in that construct, that should work out fine.

mficzel replied:
I just thought aswell that a FusionSourceCodeInterface that is implemented by `FusionSourceCode` and the `FusionSourceCodeCollection` might be a good idea. However this would have to include the method "getContextPath" which makes no sense on collection level at all and this is what this pr is about. 

So i get that the collection is just a collection.

kitsunet replied:
Doesn't need to be one interface for everything, for this specific method at least just the iteration is important, and as said, I could see a way to create an interface that allows you to iterate and only returns FusionCode objects so that at that point you can be sure to have getContextPath

mhsdesign commented:
I think eventhough its a bugfix i much rather like it to go into 8.3 - seems more like the safe bet as so many internals are touched ;)

mhsdesign commented:
Ill try to get https://github.com/neos/neos-development-collection/issues/3985 fixed too

mhsdesign committed:
Task: Minor cleanup of the FusionService/Parser/Runtime API

The goal was not to violate our @api guideline so no public @api methods are changed.

The FusionService is now partly deprecated.

Fixes #3985

mhsdesign commented:
I created a suuuuuuuuper duper hackish light weight fix for 8.0 so we can relaxed rebase this to 8.3 https://github.com/neos/neos-development-collection/pull/4056

mhsdesign committed:
LOVE: The style ci

mhsdesign committed:
Task: Better testable FusionConfigurationCache via constructor injection

mhsdesign committed:
Task: Cleanup FusionSourceCodeFactory

mhsdesign committed:
Task: Adjust naming of FusionSourceCode::fromDangerousPotentiallyDifferingSourceCodeAndContextPath

mhsdesign committed:
Task: Make $sourceCode lazy of FusionSourceCode

mhsdesign committed:
Task: Remove direct replacement for deprecated FusionService::createRuntime

mhsdesign committed:
Task: Rename FusionSourceCodeCollection::tryFromPackageRootFusion

mhsdesign commented on Neos.Neos/Tests/Unit/Domain/Service/FusionConfigurationCacheTest.php:
```diff
@@ -0,0 +1,101 @@
+        $this->mockSite = $this->getMockBuilder(Site::class)->disableOriginalConstructor()->getMock();
+        $this->mockSite->method('getSiteResourcesPackageKey')->willReturn('Some.ResourcePackage');
+
+        $this->fusionConfigurationCache = new FusionConfigurationCache(
```
constructor injection FTW!!!!

mhsdesign committed:
Task: Clean up ugly FusionSourceCodeFactoryPrototypeGeneratorTest

mhsdesign committed:
Task: Internal naming things

mficzel reviewed (approved):
Fine by reading again (not every line) ... and testing

mhsdesign committed:
Docs: FusionConfiguration

ahaeslich reviewed (approved):
Fine by reading & thanks to @mhsdesign for explaining the changes.

mficzel commented:
@mhsdesign the changes for monocle were quite minimal https://github.com/sitegeist/Sitegeist.Monocle/releases/tag/v7.10.0
